src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -133,111 +133,75 @@ private unsafe ValueTask<int> ReadAsyncInternal(Memory<byte> destination, Cancel
+                    // touch the file pointer location at all.  We will adjust it
+                    // ourselves, but only in memory. This isn't threadsafe.
+                    _filePosition += destination.Length;
                 }
-                else if (errorCode != Interop.Errors.ERROR_IO_PENDING)
-                {
-                    if (!_fileHandle.IsClosed && CanSeek)  // Update Position - It could be anywhere.
-                    {
-                        _filePosition = positionBefore;
-                    }
 
-                    valueTaskSource.ReleaseNativeResource();
-                    TryToReuse(valueTaskSource);
-
-                    if (errorCode == Interop.Errors.ERROR_HANDLE_EOF)
-                    {
-                        ThrowHelper.ThrowEndOfFileException();
-                    }
-                    else
+                // Queue an async ReadFile operation.
+                if (Interop.Kernel32.ReadFile(_fileHandle, (byte*)vts._memoryHandle.Pointer, destination.Length, IntPtr.Zero, nativeOverlapped) == 0)
+                {
+                    // The operation failed, or it's pending.
+                    int errorCode = FileStreamHelpers.GetLastWin32ErrorAndDisposeHandleIfInvalid(_fileHandle);
+                    switch (errorCode)
                     {
-                        throw Win32Marshal.GetExceptionForWin32Error(errorCode, _path);
+                        case Interop.Errors.ERROR_IO_PENDING:
+                            // Common case: IO was initiated, completion will be handled by callback.
+                            // Register for cancellation now that the operation has been initiated.
+                            vts.RegisterForCancellation(cancellationToken);
+                            break;
+
+                        case Interop.Errors.ERROR_BROKEN_PIPE:
+                            // EOF on a pipe. Callback will not be called.
+                            // We clear the overlapped status bit for this special case (failure
+                            // to do so looks like we are freeing a pending overlapped later).
+                            nativeOverlapped->InternalLow = IntPtr.Zero;
+                            vts.Dispose();
+                            return ValueTask.FromResult(0);
+
+                        default:
+                            // Error. Callback will not be called.
+                            vts.Dispose();
+                            return ValueTask.FromException<int>(HandleIOError(positionBefore, errorCode));
                     }
                 }
-                else if (cancellationToken.CanBeCanceled) // ERROR_IO_PENDING
-                {
-                    // Only once the IO is pending do we register for cancellation
-                    valueTaskSource.RegisterForCancellation(cancellationToken);
-                }
             }
-            else
+            catch
```

Comment:
this catch block seems to be redundant:
* `ERROR_BROKEN_PIPE` and `default` dispose the token on their own
* `ERROR_IO_PENDING` invokes `RegisterForCancellation` which already has exception handling logic.

IMO we should remove this catch block and extend the exception handling logic of the `RegisterForCancellation` method

Reply:
It's not redundant.  For example, what if the ReadFile throws ObjectDisposedException because the file handle was concurrently disposed?